pkg/controller/servicednsendpoint/servicednsendpoint_controller.go

```diff
@@ -0,0 +1,475 @@
+		d.queue.Forget(key)
+	} else if d.queue.NumRequeues(key) < maxRetries {
+		glog.Errorf("Error processing %s (will retry): %v", key, err)
+		// requeue the item to work on later
+		d.queue.AddRateLimited(key)
+	} else {
+		// err != nil and too many retries
+		glog.Errorf("Error processing %s (giving up): %v", key, err)
+		d.queue.Forget(key)
+		runtime.HandleError(err)
+	}
+
+	return true
+}
+
+func (d *ServiceDNSEndpointController) processItem(key string) error {
+	startTime := time.Now()
+	glog.V(4).Infof("Processing change to ServiceDNSEndpoint %s", key)
+	defer func() {
+		glog.V(4).Infof("Finished processing ServiceDNSEndpoint %q (%v)", key, time.Since(startTime))
+	}()
+
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+
+	obj, exists, err := d.serviceDNSObjectStore.GetByKey(key)
+	if err != nil {
+		return fmt.Errorf("error fetching object with key %s from store: %v", key, err)
+	}
+
+	if !exists {
+		//delete corresponding DNSEndpoint object
+		return d.client.MulticlusterdnsV1alpha1().DNSEndpoints(namespace).Delete(name, &metav1.DeleteOptions{})
+	}
+
+	dnsObject, ok := obj.(*feddnsv1a1.MultiClusterServiceDNSRecord)
+	if !ok {
+		return fmt.Errorf("recieved event for unknown object %v", obj)
+	}
+
+	dnsEndpoints, err := GetEndpointsForServiceDNSObject(dnsObject)
+	if err != nil {
+		return err
+	}
+
+	dnsEndpointObject, err := d.client.MulticlusterdnsV1alpha1().DNSEndpoints(namespace).Get(name, metav1.GetOptions{})
+	if err != nil {
+		if errors.IsNotFound(err) {
+			newDNSEndpointObject := &feddnsv1a1.DNSEndpoint{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      name,
+					Namespace: namespace,
+				},
+				Spec: feddnsv1a1.DNSEndpointSpec{
+					Endpoints: dnsEndpoints,
+				},
+			}
+
+			_, err = d.client.MulticlusterdnsV1alpha1().DNSEndpoints(namespace).Create(newDNSEndpointObject)
+		}
+		return err
+	}
+
+	// Update only if the new endpoints are not equal to the existing ones.
+	if !reflect.DeepEqual(dnsEndpointObject.Spec.Endpoints, dnsEndpoints) {
+		dnsEndpointObject.Spec.Endpoints = dnsEndpoints
+		_, err = d.client.MulticlusterdnsV1alpha1().DNSEndpoints(namespace).Update(dnsEndpointObject)
+	}
+
+	return err
+}
+
+// GetEndpointsForServiceDNSObject returns endpoint objects for each MultiClusterServiceDNSRecord object that should be processed.
+func GetEndpointsForServiceDNSObject(dnsObject *feddnsv1a1.MultiClusterServiceDNSRecord) ([]*feddnsv1a1.Endpoint, error) {
+	endpoints, err := getNamesForDNSObject(dnsObject)
+	if err != nil {
+		return nil, err
+	}
+
+	return DedupeAndMergeEndpoints(endpoints), nil
+}
+
+// Merge and remove duplicate endpoints
+func DedupeAndMergeEndpoints(endpoints []*feddnsv1a1.Endpoint) (result []*feddnsv1a1.Endpoint) {
+	// Sort endpoints by DNSName
+	sort.Slice(endpoints, func(i, j int) bool {
+		return endpoints[i].DNSName < endpoints[j].DNSName
+	})
+
+	// Remove the endpoint with no targets/ empty targets
```

Comment:
In what cases is this true? It seems like there would always be a valid string in the Targets string array as there are checks for that before inserting values.

Reply:
This may happen if there are no endpoints for a service in any zone or region for the global service. we would end up with 3 CNAME records, and the last CNAME record (global level) has empty target to point to nowhere. In such case we don't write the last CNAME record. here is the test case for it https://github.com/kubernetes-sigs/federation-v2/blob/master/pkg/controller/servicednsendpoint/servicednsendpoint_controller_test.go#L192